### PR TITLE
Add unit option for when hit points damaged changes into

### PR DESCRIPTION
**Pull request by ron-murhammer**
Addresses the first part of feature request: https://forums.triplea-game.org/topic/327/unit-option-when-damaged-change-into-different-unit-weakened-battleships

**Functional Changes**
- Added new unit option whenHitPointsDamagedChangesInto to allow units to change into a different unit type when damaged during battle

**Tested**
- Added the below example XML to TWW and tested running battles where battleships are damaged

**Example XML**
```
    <attachment name="unitAttachment" attachTo="germanBattleship" javaClass="games.strategy.triplea.attachments.UnitAttachment" type="unitType">
      <option name="movement" value="2"/>
      <option name="attack" value="7"/>
      <option name="defense" value="8"/>
      <option name="canBombard" value="true"/>
      <option name="isSea" value="true"/>
      <option name="hitPoints" value="2"/>
      <option name="bombard" value="6"/>
      <option name="requiresUnits" value="germanDocks:germanFactory"/>
      <option name="consumesUnits" value="1:germanHull"/>
      <option name="whenHitPointsDamagedChangesInto" value="1:true:germanBattleshipDamaged"/>
    </attachment>
    <attachment name="unitAttachment" attachTo="germanBattleshipDamaged" javaClass="games.strategy.triplea.attachments.UnitAttachment" type="unitType">
      <option name="movement" value="1"/>
      <option name="attack" value="3"/>
      <option name="defense" value="4"/>
      <option name="canBombard" value="true"/>
      <option name="isSea" value="true"/>
      <option name="hitPoints" value="2"/>
      <option name="bombard" value="3"/>
    </attachment>
```

### CODE REVIEW

File: src/main/java/games/strategy/triplea/attachments/UnitAttachment.java [hunk 1]
```diff
@@ -434,6 +436,46 @@ public void resetCanBeCapturedOnEnteringBy() {
...
+    }
+    return m_whenHitPointsDamagedChangesInto;
+  }
+
+  public void clearWhenHitPointsDamagedChangesInto() {
+    m_whenHitPointsDamagedChangesInto.clear();
+  }
+
+  public void resetWhenHitPointsDamagedChangesInto() {
+    m_whenHitPointsDamagedChangesInto = new LinkedHashMap<>();
```
Comment (ssoloff) [inline src/main/java/games/strategy/triplea/attachments/UnitAttachment.java]: This method uses `LinkedHashMap`, but the field initializer uses `HashMap`.  Was that intentional?

File: src/main/java/games/strategy/triplea/attachments/UnitAttachment.java [hunk 2]
```diff
@@ -434,6 +436,46 @@ public void resetCanBeCapturedOnEnteringBy() {
...
+  @GameProperty(xmlProperty = true, gameProperty = true, adds = false)
+  public void setWhenHitPointsDamagedChangesInto(final Map<Integer, Tuple<Boolean, UnitType>> value) {
+    m_whenHitPointsDamagedChangesInto = value;
+  }
+
+  /**
+   * Can remove null check and this comment for next incompatible release.
+   */
+  public Map<Integer, Tuple<Boolean, UnitType>> getWhenHitPointsDamagedChangesInto() {
+    if (m_whenHitPointsDamagedChangesInto == null) {
```
Comment (ssoloff) [inline src/main/java/games/strategy/triplea/attachments/UnitAttachment.java]: Do you need a similar `null` check in `clearWhenHitPointsDamagedChangesInto()` and `setWhenHitPointsDamagedChangesInto(String)`?  Or will those code paths not be hit after deserializing an old save game?

Comment (ron-murhammer) [reply src/main/java/games/strategy/triplea/attachments/UnitAttachment.java]: Fairly sure those shouldn't be hit.

File: src/main/java/games/strategy/triplea/attachments/UnitAttachment.java (hunk 1 above)
Comment (ron-murhammer) [reply src/main/java/games/strategy/triplea/attachments/UnitAttachment.java]: Fixed.

**Pull request merged**